docs/proposals/flexvolume-api.md

```diff
@@ -0,0 +1,352 @@
+<driver executable> probe
+```
+
+#### Provision/Create a volume:
+This call creates a volume. It is executed from Controller-manager.
+
+###### New API Types:
+```go
+Type FlexVolumeStatus string
+
+type FlexVolume struct {
+    unversioned.TypeMeta
+    v1.ObjectMeta
+    Spec api.FlexVolumeSource
+    Status FlexVolumeStatus
+}
+```
+
+###### Request:
+```
+FlexVolume
+```
+
+###### Response:
+On success:
+```
+api.FlexVolume
+```
+On failure:
+```
+non-zero exit code & errors.StatusError
+```
+
+###### Call out:
+```
+<driver executable> create <json encoded api.FlexVolume>
+```
+
+#### Delete a volume:
+This call deletes a volume. It is executed from Controller-manager.
+
+###### Request:
+```
+FlexVolume
+```
+
+###### Response:
+On success:
+```
+0 exit code
+```
+On failure:
+```
+non-zero exit code & errors.StatusError
+```
+
+###### Call out:
+```
+<driver executable> delete <json encoded api.FlexVolume>
+```
+
+#### Attach a volume:
+This call attaches a volume to a remote host or local host. It can be executed from Controller-manager/Kubelet depending on whether Controller-attach-detach is enabled or not.. This is only valid for drivers which support attach & detach.
+
+The following additional options are passed to the plugin as part of options map in FlexVolumeSource.
+```
+“kubernetes.io/host”=”<host-name>”
+“kubernetes.io/mountpath”=”<mount path>”
+```
+
+###### New API Types:
+```go
+type FlexVolumeAttachment struct {
+    unversioned.TypeMeta
+    v1.ObjectMeta
+    Host string
+    Device string
+}
+```
+
+###### Request:
+```
+FlexVolume
+```
+
+###### Response:
+On success:
+```
+FlexVolumeAttachment
+```
+On failure:
+```
+non-zero exit code & errors.StatusError
+```
+
+###### Call out:
+```
+<driver executable> attach <json encoded FlexVolume>
+```
+
+###### Call out sync: (Wait for attach):
+```
+<driver executable> attach <json encoded FlexVolume> sync
+```
+
+#### Detach a volume:
+This call detaches a volume to a remote host or local host. It can be executed from Controller-manager/Kubelet depending on whether Controller-attach-detach is enabled or not.. This call only valid for drivers which support attach & detach.
+
+The following additional options are passed to the plugin as part of options map in FlexVolumeSource.
+```
+“kubernetes.io/host”=”<host-name>”
+```
+
+###### Request:
+```
+FlexVolumeAttachment
+```
+
+###### Response:
+On success:
+```
+0 exit code
+```
+On failure:
+```
+non-zero exit code & errors.StatusError
+```
+
+###### Call out:
+```
+<driver executable> detach <json encoded FlexVolumeAttachment>
+```
+
+###### Call out sync: (Wait for detach)
+```
+<driver executable> detach <json encoded FlexVolumeAttachment> sync
+```
+
+#### Mount a volume:
+This call mounts a volume on the node. It is executed from Kubelet. This is only valid for plugins which do not support attach/detach to a node. Example: NFS/CIFS. For volumes which support attach/detach, this will use the default logic to bind mount.
```

Comment:
Yes. On second thought separated it out. With the new custommount option driver can dictate it. Will update the proposal.